File: forms/medical-operation-note/front-end-with-svelte/src/lib/engine/anaesthetic-event-rules.ts
import type { CompositeRisk, FiredRule, OperationNote } from './types.js';

/**
 * Parse intra-operative anaesthetic events from
 * `anaesthesia.events` (comma-separated tokens). Each known major event
 * is a high-priority fired rule and lifts composite risk.
 *
 * Tokens (lowercased, trimmed):
 *   failed-intubation, awareness, anaphylaxis, malignant-hyperthermia,
 *   sux-apnoea, hypotension, hypertension, bronchospasm, laryngospasm
 *
 * Major events → composite risk ≥ High-risk.
 * Minor events (hypotension, hypertension, bronchospasm, laryngospasm)
 * → composite risk ≥ Complicated.
 */
const MAJOR = new Set([
  'failed-intubation',
  'awareness',
  'anaphylaxis',
  'malignant-hyperthermia',
  'sux-apnoea',
]);
const MINOR = new Set(['hypotension', 'hypertension', 'bronchospasm', 'laryngospasm']);

export function applyAnaestheticEventRules(data: OperationNote): {
  risk: CompositeRisk;
  firedRules: FiredRule[];
  majorTokens: string[];
  minorTokens: string[];
} {
  const raw = data.anaesthesia.events || '';
  const tokens = raw
    .split(',')
    .map((t) => t.trim().toLowerCase())
    .filter((t) => t.length > 0);

  const majorTokens: string[] = [];
  const minorTokens: string[] = [];
  const firedRules: FiredRule[] = [];

  for (const t of tokens) {
    if (MAJOR.has(t)) {
      majorTokens.push(t);
      firedRules.push({
        ruleId: `R-AN-MAJOR-${t.toUpperCase()}`,
        instrument: 'anaesthetic-event',
        grade: 'major',
        category: 'anaesthesia',
        description: `Major anaesthetic event: ${t}`,
      });
    } else if (MINOR.has(t)) {
      minorTokens.push(t);
      firedRules.push({
        ruleId: `R-AN-MINOR-${t.toUpperCase()}`,
        instrument: 'anaesthetic-event',
        grade: 'minor',
        category: 'anaesthesia',
        description: `Minor anaesthetic event: ${t}`,
      });
    }
  }

  let risk: CompositeRisk = 'routine';
  if (majorTokens.length > 0) risk = 'high-risk';
  else if (minorTokens.length > 0) risk = 'complicated';

  return { risk, firedRules, majorTokens, minorTokens };
}

File: forms/medical-operation-note/back-end-with-loco/src/medical_operation_note/engine/anaesthetic_event_rules.rs
//! Anaesthetic-event rules.
//!
//! Failed intubation, awareness, anaphylaxis, malignant hyperthermia and
//! suxamethonium apnoea are all high-priority. Minor events keep the
//! composite at Complicated.

use super::types::{CompositeRisk, FiredRule};

const HIGH_SEVERITY: [&str; 5] = [
    "failed-intubation",
    "awareness",
    "anaphylaxis",
    "malignant-hyperthermia",
    "suxamethonium-apnoea",
];

/// Evaluate.
#[must_use]
pub fn evaluate(event: &str) -> (CompositeRisk, Option<FiredRule>) {
    if event.is_empty() || event == "none" {
        return (CompositeRisk::Routine, None);
    }
    if HIGH_SEVERITY.contains(&event) {
        return (
            CompositeRisk::HighRisk,
            Some(FiredRule {
                rule_id: format!("anaesthetic.{event}"),
                instrument: "anaesthetic".into(),
                band: "high-risk".into(),
                category: "anaesthetic-incident".into(),
                description: format!("Anaesthetic incident: {event}"),
            }),
        );
    }
    (
        CompositeRisk::Complicated,
        Some(FiredRule {
            rule_id: format!("anaesthetic.{event}"),
            instrument: "anaesthetic".into(),
            band: "complicated".into(),
            category: "anaesthetic-event".into(),
            description: format!("Minor anaesthetic event: {event}"),
        }),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_routine() {
        assert_eq!(evaluate("").0, CompositeRisk::Routine);
        assert_eq!(evaluate("none").0, CompositeRisk::Routine);
    }

    #[test]
    fn minor_complicated() {
        assert_eq!(evaluate("hypotension").0, CompositeRisk::Complicated);
    }

    #[test]
    fn major_high_risk() {
        assert_eq!(evaluate("failed-intubation").0, CompositeRisk::HighRisk);
        assert_eq!(evaluate("anaphylaxis").0, CompositeRisk::HighRisk);
        assert_eq!(evaluate("malignant-hyperthermia").0, CompositeRisk::HighRisk);
    }
}
